packages/front-end/components/Experiment/AnalysisForm.tsx

```diff
@@ -243,17 +249,56 @@ const AnalysisForm: FC<{
         labelClassName="font-weight-bold"
         value={datasource?.id || ""}
         onChange={(newDatasource) => {
-          if (datasource && newDatasource !== datasource?.id) {
+          // Don't do anything if the selected one doesn't change
+          if (newDatasource === (datasource?.id || "")) {
+            return;
+          }
+
+          const segment = form.watch("segment");
+          const activationMetric = form.watch("activationMetric");
+          const metrics = form.watch("metrics");
+          const guardrails = form.watch("guardrails");
+
+          form.setValue("exposureQueryId", "");
+
+          if (
+            segment &&
+            getSegmentById(segment)?.datasource !== newDatasource
+          ) {
             form.setValue("segment", "");
+          }
+
+          if (
+            activationMetric &&
+            getMetricById(activationMetric)?.datasource !== newDatasource
+          ) {
             form.setValue("activationMetric", "");
-            form.setValue("exposureQueryId", "");
           }
+
+          form.setValue(
+            "metrics",
+            metrics.filter(
+              (m) => getMetricById(m)?.datasource === newDatasource
+            )
+          );
+          form.setValue(
+            "guardrails",
+            guardrails.filter(
+              (m) => getMetricById(m)?.datasource === newDatasource
+            )
+          );
+
           form.setValue("datasource", newDatasource);
         }}
-        options={datasources.map((d) => ({
-          value: d.id,
-          label: `${d.name}${d.description ? ` — ${d.description}` : ""}`,
-        }))}
+        options={datasources
+          .filter(
+            (ds) =>
+              !ds.projects || ds.projects.includes(experiment.project || "")
```

Comment:
The coalescing seems slightly risky. What if `ds.projects === []`

Reply:
yeah, that's a good point.  I'll abstract this logic to a helper function and make it more robust